### PR TITLE
Added proper headers for artifacts downloads

### PR DESCRIPTION
**Pull request by bali0019**
Signed-off-by: Jas Bali <bali0019@gmail.com>

## Related Issues/PRs

 #7548

## What changes are proposed in this pull request?

For artifacts downloads API, ensured proper headers are added and mime type is correctly resolved.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/server/handlers.py [hunk 1]
```diff
@@ -88,6 +89,7 @@
 _model_registry_store = None
 _artifact_repo = None
 STATIC_PREFIX_ENV_VAR = "_MLFLOW_STATIC_PREFIX"
+_mime_type = mimetypes.MimeTypes()
```
Comment (BenWilson2) [inline mlflow/server/handlers.py:92]: Do we need this? Isn't the func `guess_type()` in the `__init__()` of the `mimetypes` package?
Could we just import this function directly `from mimetypes import guess_type`?

File: mlflow/server/handlers.py [hunk 2]
```diff
@@ -495,6 +507,20 @@ def wrapper(*args, **kwargs):
...
 
 
+def _guess_mime_type(filename):
+    extension = os.path.splitext(filename)[-1].replace(".", "")
+    # for MLmodel/mlproject with no extensions
+    if extension == "":
+        extension = filename
+    if extension in _TEXT_EXTENSIONS:
+        return "text/plain"
+    mime_type = _mime_type.guess_type(filename)[0]
```
Comment (BenWilson2) [inline mlflow/server/handlers.py:517]:
```suggestion
    mime_type, _ = _mime_type.guess_type(filename)
```

File: tests/tracking/test_mlflow_artifacts.py [hunk 3]
```diff
@@ -321,3 +326,41 @@ def test_rest_get_model_version_artifact_api_proxied_artifact_root(artifacts_ser
...
+    with mlflow.start_run() as run:
+        mlflow.log_artifact(test_file)
+    artifact_response = requests.get(
+        url=f"{url}/get-artifact", params={"run_id": run.info.run_id, "path": filename}
+    )
+    artifact_response.raise_for_status()
+    _, params = cgi.parse_header(artifact_response.headers["Content-Disposition"])
+    assert params["filename"] == filename
+    assert artifact_response.headers["Content-Type"] == expected_mime_type
+    assert artifact_response.headers["X-Content-Type-Options"] == "nosniff"
```
Comment (BenWilson2) [inline tests/tracking/test_mlflow_artifacts.py:366]: Can we add a test where we write a particular file with an extension, but in the request we explicitly add a different mimetype header to simulate nefarious behavior and ensure that the server returns the correct mimetype and ignores the requested type?

File: mlflow/server/handlers.py [hunk 4]
```diff
@@ -495,6 +506,20 @@ def wrapper(*args, **kwargs):
 ]
 
 
+def _guess_mime_type(filename):
+    extension = os.path.splitext(filename)[-1].replace(".", "")
+    # for MLmodel/mlproject with no extensions
+    if extension == "":
+        extension = filename
```
Comment (BenWilson2) [inline mlflow/server/handlers.py:513]: is there something odd happening with windows on this logic assignment where it's not recognizing the empty extension type?

File: mlflow/server/handlers.py [hunk 5]
```diff
@@ -495,6 +507,21 @@ def wrapper(*args, **kwargs):
 ]
 
 
+def _guess_mime_type(file_path):
+    filename = posixpath.basename(pathlib.Path(file_path).as_posix())
+    extension = os.path.splitext(filename)[-1].replace(".", "")
```
Comment (harupy) [inline mlflow/server/handlers.py:512]:
```suggestion
    extension = pathlib.Path(file_path).suffix.replace(".", "")
```

I haven't tested but this should work.

File: tests/server/test_handlers.py [hunk 6]
```diff
@@ -641,3 +643,39 @@ def test_delete_model_version_tag(mock_get_request_message, mock_model_registry_
...
+        ("c.txt", "text/plain"),
+        ("c.pkl", "application/octet-stream"),
+        ("C:\\a\\b\\c.pkl", "application/octet-stream"),
+        ("C:\\a\\b\\c.png", "image/png"),
+        ("C:\\a\\b\\c.pdf", "application/pdf"),
+        ("C:\\a\\b\\MLmodel", "text/plain"),
+        ("C:\\a\\b\\mlproject", "text/plain"),
+    ],
+)
+def test_guess_mime_type_on_windows(file_path, expected_mime_type):
```
Comment (harupy) [inline tests/server/test_handlers.py:680]:
```suggestion
@pytest.mark.parametrize(
    ("file_path", "expected_mime_type"),
    [
        ("/a/b/c.txt", "text/plain"),
        ("c.txt", "text/plain"),
        ("c.pkl", "application/octet-stream"),
        ("/a/b/c.pkl", "application/octet-stream"),
        ("/a/b/c.png", "image/png"),
        ("/a/b/c.pdf", "application/pdf"),
        ("/a/b/MLmodel", "text/plain"),
        ("/a/b/mlproject", "text/plain"),
        ("C:\\a\\b\\c.txt", "text/plain"),
        ("c.txt", "text/plain"),
        ("c.pkl", "application/octet-stream"),
        ("C:\\a\\b\\c.pkl", "application/octet-stream"),
        ("C:\\a\\b\\c.png", "image/png"),
        ("C:\\a\\b\\c.pdf", "application/pdf"),
        ("C:\\a\\b\\MLmodel", "text/plain"),
        ("C:\\a\\b\\mlproject", "text/plain"),
    ],
)
def test_guess_mime_type(file_path, expected_mime_type):
```

I think we can merge these tests.

Comment (bali0019) [reply tests/server/test_handlers.py:680]: I tested this with the above change `filename = pathlib.Path(file_path).name`, on unix systems that returns the entire path as name.
input: `pathlib.Path("C:\\a\\b\\MLmodel").name`
output: `C:\\a\\b\\MLmodel`

I think we need to separate those tests by OS here to assert that the artifact path resolves to a correct mime type.

Comment (harupy) [reply tests/server/test_handlers.py:680]: Got it, thanks for checking!

**Review by harupy — APPROVED** [review]
LGTM!

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @bali0019 ! Super slick PR!

**Pull request merged**